core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -70,35 +88,164 @@
+    final CompletableFuture<T> future = new CompletableFuture<>();
+    future.completeExceptionally(e);
+    return future;
+  }
+
+  @Override
+  public void initialize(RaftServer server, RaftGroupId groupId,
+      RaftStorage raftStorage) throws IOException {
+    getLifeCycle().startAndTransition(() -> {
+      super.initialize(server, groupId, raftStorage);
+      mRaftGroupId = groupId;
+      mStorage.init(raftStorage);
+      loadSnapshot(mStorage.getLatestSnapshot());
+    });
+  }
+
+  @Override
+  public void reinitialize() throws IOException {
+    LOG.info("Reinitializing state machine.");
+    mStorage.loadLatestSnapshot();
+    loadSnapshot(mStorage.getLatestSnapshot());
+    unpause();
+  }
+
+  private void loadSnapshot(SingleFileSnapshotInfo snapshot) throws IOException {
+    if (snapshot == null) {
+      LOG.info("No snapshot to load");
+      return;
+    }
+    LOG.info("Loading Snapshot {}", snapshot);
+    final File snapshotFile = snapshot.getFile().getPath().toFile();
+    if (!snapshotFile.exists()) {
+      LOG.error("The snapshot {} does not exist", snapshot);
+      throw new FileNotFoundException(
+          String.format("The snapshot file %s does not exist", snapshotFile.getPath()));
+    }
+    try {
+      resetState();
+      setLastAppliedTermIndex(snapshot.getTermIndex());
+      install(snapshotFile);
+    } catch (Exception e) {
+      LOG.error("Failed to load snapshot {}", snapshot, e);
+      throw e;
+    }
+  }
+
+  @Override
+  public long takeSnapshot() {
+    if (mIsLeader) {
+      // TODO(feng): if secondary master has a more recent snapshot, install it
+      return RaftLog.INVALID_LOG_INDEX;
+    } else {
+      return takeLocalSnapshot();
+    }
+  }
+
+  @Override
+  public SnapshotInfo getLatestSnapshot() {
+    return mStorage.getLatestSnapshot();
+  }
+
+  @Override
+  public StateMachineStorage getStateMachineStorage() {
+    return mStorage;
+  }
+
+  @Override
+  public CompletableFuture<Message> query(Message request) {
+    LOG.info("Received query: {}", request);
+    return super.query(request);
+  }
+
+  @Override
+  public void close() {
+    mClosed = true;
+  }
+
+  @Override
+  public CompletableFuture<Message> applyTransaction(TransactionContext trx) {
+    try {
+      applyJournalEntryCommand(trx);
+      return super.applyTransaction(trx);
+    } catch (Exception e) {
+      return completeExceptionally(e);
+    }
+  }
+
+  @Override
+  public void notifyNotLeader(Collection<TransactionContext> pendingEntries) {
+    mIsLeader = false;
+    mJournalSystem.notifyLeadershipStateChanged(false);
+  }
+
+  @Override
+  public void pause() {
+    getLifeCycle().transition(LifeCycle.State.PAUSING);
+    try {
+      if (!mJournalApplier.isSuspended()) {
+        suspend();
+      }
+    } catch (IOException e) {
+      LOG.error("State machine pause failed", e);
+      throw new IllegalStateException(e);
```

Comment:
Do we need to throw and log? Can we have this message include the message above?

Reply:
Removed the logging message.